In [1]:
import pandas as pd
import numpy as np
import os 

Uploading products.csv and orders.csv

In [106]:
path = '/Users/satoruteshima/Documents/CareerFoundry/Data Immersion 4/08-2023 Instacart Basket Analysis'

In [5]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', '01 Origina Data', 'products.csv'), index_col = False)

In [124]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', '02 Prepared Data', 'orders_wrangled.csv'), index_col = False)

The df.describe() function returns descriptive statistics for the numeric values in your dataframe.

In [125]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_last_order  
count        3.421083e+06       3.421083e+06           3.214874e+06  
mean         2.776219e+00       1.345202e+01           1.111484e+01  
std          2.046829e+00       4.226088e+00           9.206737e+00  
min          0.000000e+00       0.000000e+00           0.000000e+00  
25%          1.000000e+00       1.000000e+01           4.000000e+00  
50%          3.000000e+00       1.300000e+01           7.000000e+00  
75%          5.000000e+00       1.600000e+01           1.500000e+01  
max          6.000000e+00       2.300000e+01           3.000000e+01

## Mixed Data Type


In [12]:
#creating test dataframe
df_test = pd.DataFrame()

In [13]:
#create a mixed data column
df_test ['mix'] = ['a', 'b', '1', True]


In [14]:
#check your test dataframe
df_test.head()

mix
0     a
1     b
2     1
3  True

The function for checking whether a dataframe contains any mixed-type columns is as follows:

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

In [1]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

NameError: name 'df_test' is not defined

In [19]:
#astype will assign a datatype 
df_test['mix'] = df_test['mix'].astype('str')

## Missing Values


### Finding Missing Values


In [22]:
#isnull() function  function is used to find missing observations(cells)

df_prods.isnull().sum()


product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

Results above shows that there are 16 missing values (16 Trues are summed up) 

In [23]:
#Create a new dataframe, df_nan, containing only those values within the "product_name" column that meet the condition isnull() = True

df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [24]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

Now that you’ve identified the rows with missing values, let’s take a look at how you can address them in order to prepare your data for analysis.

### Addressing missing values

In [26]:
df_prods.shape

(49693, 5)

In [27]:
#create a new dataframe called df_prods_clean

df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]


In [28]:
df_prods_clean.shape

(49677, 5)

Alternative way to drop missing values 

df_prods.dropna(inplace = True)

to drop only the NaNs from a particular column

df_prods.dropna(subset = [‘product_name’], inplace = True)


## Duplicates

In [31]:
df_prods_clean[df_prods_clean.duplicated()]

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

 The following command will look for full duplicates within your dataframe:


The duplicated() function is what identifies duplicate rows.

In [42]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [45]:
#drop_duplicates () drops the duplicate data from the dataframe 

df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [44]:
#comparing the original and the prepared data
df_prods_clean.shape

(49677, 5)

In [47]:
df_prods_clean_no_dups.shape

(49672, 5)

In [108]:
#export 
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','02 Prepared Data', 'products_checked.csv'))

# TASKS

#### Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.

In [126]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_last_order  
count        3.421083e+06       3.421083e+06           3.214874e+06  
mean         2.776219e+00       1.345202e+01           1.111484e+01  
std          2.046829e+00       4.226088e+00           9.206737e+00  
min          0.000000e+00       0.000000e+00           0.000000e+00  
25%          1.000000e+00       1.000000e+01           4.000000e+00  
50%          3.000000e+00       1.300000e+01           7.000000e+00  
75%          5.000000e+00       1.600000e+01           1.500000e+01  
max          6.000000e+00       2.300000e+01           3.000000e+01

Insights

- The max value of order_number, order_hour_of_day, days_since_last_order do not seem to match the actual data. 
- The count value of order_number, order_hour_of_dat, days_since_last_order do not seem to match the actual data. 
-  The standard deviation of days_since_last order is 9.2. That means it cointains a lot of outliers.

#### Check for mixed-type data in your df_ords dataframe.


In [129]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

No mixed type data

### Run a check for missing values in your df_ords dataframe. In a markdown cell, report your findings and propose an explanation for any missing values you find.

In [130]:
df_ords.isnull().sum()

Unnamed: 0                    0
order_id                      0
user_id                       0
order_number                  0
orders_day_of_week            0
order_hour_of_day             0
days_since_last_order    206209
dtype: int64

Findings

- days_since_last_order contain 206209 missing values 


In [157]:
#locating the missing value and assigning it to a new dataframe
df_ords_nan = df_ords[df_ords['days_since_last_order'].isnull() == True]


In [139]:
#checking data

df_ords_nan.describe()

Unnamed: 0      order_id        user_id  order_number  \
count  2.062090e+05  2.062090e+05  206209.000000      206209.0   
mean   1.712637e+06  1.708462e+06  103105.000000           1.0   
std    9.874816e+05  9.881299e+05   59527.555167           0.0   
min    0.000000e+00  2.000000e+01       1.000000           1.0   
25%    8.579730e+05  8.507300e+05   51553.000000           1.0   
50%    1.717068e+06  1.706246e+06  103105.000000           1.0   
75%    2.570518e+06  2.564292e+06  154657.000000           1.0   
max    3.421069e+06  3.421081e+06  206209.000000           1.0   

       orders_day_of_week  order_hour_of_day  days_since_last_order  
count       206209.000000      206209.000000                    0.0  
mean             2.754118          13.626597                    NaN  
std              2.076205           4.223769                    NaN  
min              0.000000           0.000000                    NaN  
25%              1.000000          11.000000                    NaN  
50%              3.000000          14.000000                    NaN  
75%              5.000000          17.000000                    NaN  
max              6.000000          23.000000                    NaN

In [160]:
#checking the number of values 
df_ords_nan['order_number'].value_counts()

1    206209
Name: order_number, dtype: int64

It seems the first order is treated as a missing value. 

- The std is over 2.5 therefore mean can't be trusted.  Plus, it is the first order from all users that are not recorded. 
- The records of users first order is a valuable data so I will not drop or filter it. 
- The solution is to leave it as it is. 

In [166]:
df_ords_clean = df_ords

In [178]:
#add a new column to record the first order from a new customer 
df_ords_clean['new_customer'] = df_ords['days_since_last_order'].isnull() == True

In [175]:
df_ords_clean

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
1                 1   2398795        1             2                   3   
2                 2    473747        1             3                   3   
3                 3   2254736        1             4                   4   
4                 4    431534        1             5                   4   
...             ...       ...      ...           ...                 ...   
3421078     3421078   2266710   206209            10                   5   
3421079     3421079   1854736   206209            11                   4   
3421080     3421080    626363   206209            12                   1   
3421081     3421081   2977660   206209            13                   1   
3421082     3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_last_order  new_customer  
0                        8                    NaN          True  
1                        7                   15.0         False  
2                       12                   21.0         False  
3                        7                   29.0         False  
4                       15                   28.0         False  
...                    ...                    ...           ...  
3421078                 18                   29.0         False  
3421079                 10                   30.0         False  
3421080                 12                   18.0         False  
3421081                 12                    7.0         False  
3421082                 14                   30.0         False  

[3421083 rows x 8 columns]

### Run a check for duplicate values in your df_ords data. In a markdown cell, report your findings and propose an explanation for any duplicate values you find.

In [104]:
df_ords[df_ords.duplicated()]

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_last_order]
Index: []

No duplicates. 

In [176]:
df_ords_clean.to_csv(os.path.join(path, '02 Data','02 Prepared Data', 'orders_clean_checked.csv'))